### PR TITLE
Enhance XEN VM state check

### PR DESCRIPTION
**Pull request by RoyCai7**
poo#https://progress.opensuse.org/issues/127805
VM state is not stable, enhance it.

- Related ticket: https://progress.opensuse.org/issues/127805
- Needles: n/A
- Verification run: [XEN guest migration](http://openqa.qam.suse.cz/tests/overview?build=:S:M:28650:295434:64bit-guest-mi-ipmi-group2:12-SP5:g_m_test:&distri=sle&version=12-SP5&groupid=160), [KVM guest migraion](http://openqa.qam.suse.cz/tests/overview?build=:S:M:28650:295434:64bit-guest-mi-ipmi:12-SP5:g_m_test:&distri=sle&version=12-SP5&groupid=149)

### CODE REVIEW

**Comment by RoyCai7** [comment]
@waynechen55 @alice-suse @nanzhg @guoxuguang @tbaev @Julie-CAO @varunkojha @tonyyuan1

**Comment by waynechen55** [comment]
I am doing some verification run for test suites on OSD with your branch.

**Comment by RoyCai7** [comment]
> I am doing some verification run for test suites on OSD with your branch. @RoyCai7

I verified.

**Comment by Julie-CAO** [comment]
I agree with your PR that test is supposed to be aborted when guests are not avaible.

I know your PR did not touch this part, but I'd like to know if the guest is in shutdown state, will its journals still be checked?  And another question, does a xen guest have a '-r' state in 'xl list'? because I noticed a '-r' bit but am not sure if a domU can be in this state.

**Comment by waynechen55** [comment]
> I agree with your PR that test is supposed to be aborted when guests are not avaible.
> 
> I know your PR did not touch this part, but I'd like to know if the guest is in shutdown state, will its journals still be checked? And another question, does a xen guest have a '-r' state in 'xl list'? because I noticed a '-r' bit but am not sure if a domU can be in this state.

I think powered off guest will not be shown up by ```xl list```, it can only been seen by ```virsh list```.

**Comment by waynechen55** [comment]
> > I am doing some verification run for test suites on OSD with your branch. @RoyCai7
> 
> I verified.

Why ont put them in ```verification run``` ?

**Comment by RoyCai7** [comment]
> poo#https://progress.opensuse.org/issues/127805 VM state is not stable, enhance it.
> 
> * Related ticket: https://progress.opensuse.org/issues/127805
> * Needles: n/A
> * Verification run: http://openqa.qam.suse.cz/tests/52710#step/parallel_guest_migration_destination/1873

it's there.
http://openqa.qam.suse.cz/tests/52710#step/parallel_guest_migration_destination/1873

**Comment by waynechen55** [comment]
> > poo#https://progress.opensuse.org/issues/127805 VM state is not stable, enhance it.
> > 
> > * Related ticket: https://progress.opensuse.org/issues/127805
> > * Needles: n/A
> > * Verification run: http://openqa.qam.suse.cz/tests/52710#step/parallel_guest_migration_destination/1873
> 
> it's there. http://openqa.qam.suse.cz/tests/52710#step/parallel_guest_migration_destination/1873

I mean verification run for test suites [here](https://openqa.suse.de/tests/overview?distri=sle&version=15-SP5&build=93.2&groupid=263)

**Comment by RoyCai7** [comment]
> > > poo#https://progress.opensuse.org/issues/127805 VM state is not stable, enhance it.
> > > 
> > > * Related ticket: https://progress.opensuse.org/issues/127805
> > > * Needles: n/A
> > > * Verification run: http://openqa.qam.suse.cz/tests/52710#step/parallel_guest_migration_destination/1873
> > 
> > 
> > it's there. http://openqa.qam.suse.cz/tests/52710#step/parallel_guest_migration_destination/1873
> 
> I mean verification run for test suites [here](https://openqa.suse.de/tests/overview?distri=sle&version=15-SP5&build=93.2&groupid=263).

all right, I am just focus on MU test. please paste your verification run.

**Comment by waynechen55** [comment]
> > > > poo#https://progress.opensuse.org/issues/127805 VM state is not stable, enhance it.
> > > > 
> > > > * Related ticket: https://progress.opensuse.org/issues/127805
> > > > * Needles: n/A
> > > > * Verification run: http://openqa.qam.suse.cz/tests/52710#step/parallel_guest_migration_destination/1873
> > > 
> > > 
> > > it's there. http://openqa.qam.suse.cz/tests/52710#step/parallel_guest_migration_destination/1873
> > 
> > 
> > I mean verification run for test suites [here](https://openqa.suse.de/tests/overview?distri=sle&version=15-SP5&build=93.2&groupid=263).
> 
> all right, I am just focus on MU test. please paste your verification run.

They are still in progress.

**Comment by Julie-CAO** [comment]
> > I agree with your PR that test is supposed to be aborted when guests are not available.
> > I know your PR did not touch this part, but I'd like to know if the guest is in shutdown state, will its journals still be checked? And another question, does a xen guest have a '-r' state in 'xl list'? because I noticed a '-r' bit but am not sure if a domU can be in this state.
> 
> I think powered off guest will not be shown up by `xl list`, it can only been seen by `virsh list`.

so based on current logic in test script, we will still check the journals of a shutdown guest. It does not make sense.

```
validate_script_output("virsh domstate $vm", sub { /running/ }, proceed_on_failure => 1) if (script_run("virsh list --all | grep \"$vm \"") == 0);
script_retry("xl list $vm | grep \"\\-b\\-\\-\\-\\-\"", delay => 10, retry => 30, die => 1) if (is_xen_host and script_run("xl list $vm") == 0);
my $failures = check_failures_in_journal($vm);
```

**Comment by waynechen55** [comment]
> 

If you think this part should be enhanced further, could you help come up with some solid solution ? So @RoyCai7 may help put it into this pull request as well.

**Comment by waynechen55** [comment]
> > > I agree with your PR that test is supposed to be aborted when guests are not available.
> > > I know your PR did not touch this part, but I'd like to know if the guest is in shutdown state, will its journals still be checked? And another question, does a xen guest have a '-r' state in 'xl list'? because I noticed a '-r' bit but am not sure if a domU can be in this state.
> > 
> > 
> > I think powered off guest will not be shown up by `xl list`, it can only been seen by `virsh list`.
> 
> so based on current logic in test script, we will still check the journals of a shutdown guest. It does not make sense.
> 
> ```
> validate_script_output("virsh domstate $vm", sub { /running/ }, proceed_on_failure => 1) if (script_run("virsh list --all | grep \"$vm \"") == 0);
> script_retry("xl list $vm | grep \"\\-b\\-\\-\\-\\-\"", delay => 10, retry => 30, die => 1) if (is_xen_host and script_run("xl list $vm") == 0);
> my $failures = check_failures_in_journal($vm);
> ```

If you think this part should be enhanced further, could you help come up with some solid solution ? So @RoyCai7 may help put it into this pull request as well.

**Comment by Julie-CAO** [comment]
> > > > I agree with your PR that test is supposed to be aborted when guests are not available.
> > > > I know your PR did not touch this part, but I'd like to know if the guest is in shutdown state, will its journals still be checked? And another question, does a xen guest have a '-r' state in 'xl list'? because I noticed a '-r' bit but am not sure if a domU can be in this state.
> > > 
> > > 
> > > I think powered off guest will not be shown up by `xl list`, it can only been seen by `virsh list`.
> > 
> > 
> > so based on current logic in test script, we will still check the journals of a shutdown guest. It does not make sense.
> > ```
> > validate_script_output("virsh domstate $vm", sub { /running/ }, proceed_on_failure => 1) if (script_run("virsh list --all | grep \"$vm \"") == 0);
> > script_retry("xl list $vm | grep \"\\-b\\-\\-\\-\\-\"", delay => 10, retry => 30, die => 1) if (is_xen_host and script_run("xl list $vm") == 0);
> > my $failures = check_failures_in_journal($vm);
> > ```
> 
> If you think this part should be enhanced further, could you help come up with some solid solution ? So @RoyCai7 may help put it into this pull request as well.

I just think that it does not make sense to try to ssh to a shutdown guest to check any logs. Aborting the test or just returning the guest state instead of looking into its journals is smarter.

**Comment by RoyCai7** [comment]
How about this solution:
```
validate_script_output("virsh domstate $vm", sub { /running/ }, proceed_on_failure => 1) if (script_run("virsh list --all | grep \"$vm \"") == 0);
if (is_xen_host and script_run("xl list $vm") == 0) {
    my $retry_count = 0;
    while ($retry_count < 4) {
        script_retry("xl list $vm | grep \"-b----\"", delay => 10, retry => 1, die => 0);
        $retry_count++;
    }
    validate_script_output("xl list $vm", sub { /-b----/ }, proceed_on_failure => 1);
}
my $failures = check_failures_in_journal($vm);
```
I check about 30s later, VM must become stable "-b-----".
It will solve VM stable issue, also solve the log check when shutdown VM. 
@Julie-CAO

**Comment by Julie-CAO** [comment]
> How about this solution:

> I check about 30s later, VM must become stable "-b-----". It will solve VM stable issue, also solve the log check when shutdown VM. @Julie-CAO

"-r" or "-b" both flags that a guest is running or active. Usually when a guest is booting, it is '-r'; when it boots up successfully and no other input or task running, it changes to '-b'. So I think it makes sense you wait for a well booted guest by checking this bit. It is a trick. 

```
           r - running
               The domain is currently running on a CPU.

           b - blocked
               The domain is blocked, and not running or runnable.  This can be because the domain is waiting on IO
               (a traditional wait state) or has gone to sleep because there was nothing else for it to do.
```

But I don't get why you said "also solve the log check when shutdown VM."?? What I talked previously is that we check the journal logs for running guests only, we should not check logs for a guest in shutdown state, This is meaningless and a waste of time.

**Comment by RoyCai7** [comment]
> also solve the log check when shutdown VM.

for this, after some time, VM is alive, not shutdown. So do not need check VM state to check log. this is my thought.
we have the same understanding.

**Comment by Julie-CAO** [comment]
> > also solve the log check when shutdown VM.
> 
> for this, after some time, VM is alive, not shutdown. So do not need check VM state to check log. this is my thought. I think we have the same understanding.

No, it is not what I meant.  I mean during tests, some guests might fail to boot or crash due to some reasons or failures. For these type of guests, we need not check logs from inside of them. We only look into logs for a living guest.

**Comment by waynechen55** [comment]
@RoyCai7  @Julie-CAO  I think we can drop ```validate_script_output```. Following is my proposal:
```
diff --git a/lib/virt_autotest/utils.pm b/lib/virt_autotest/utils.pm
index 99405b5c3..7b219b33f 100644
--- a/lib/virt_autotest/utils.pm
+++ b/lib/virt_autotest/utils.pm
@@ -201,12 +201,14 @@ sub check_guest_health {
     return unless is_x86_64 and ($vm =~ /sle|alp/i);
 
     #check if guest is still alive
-    validate_script_output("virsh domstate $vm", sub { /running/ }, proceed_on_failure => 1) if (script_run("virsh list --all | grep \"$vm \"") == 0);
+    my $vmstate = "nok";
+    my $failures = "";
+    $vmstate = "ok" if (script_run("virsh list --all | grep \"$vm \" | grep running") == 0);
     if (is_xen_host and script_run("xl list $vm") == 0) {
-        script_retry("xl list $vm | grep \"\\-b\\-\\-\\-\\-\"", delay => 10, retry => 30, die => 0);
-        validate_script_output("xl list $vm", sub { /-b----/ }, proceed_on_failure => 1);
+        script_retry("xl list $vm | grep \"-b----\"", delay => 5, retry => 7, die => 0);
+       $vmstate = "ok" if script_run("xl list $vm | grep \"-b----\"");
     }
-    my $failures = check_failures_in_journal($vm);
+    $failures = check_failures_in_journal($vm) if ($vmstate eq "ok");
     return 'fail' if $failures;
     record_info("Healthy guest!", "$vm looks good so far!");
     return 'pass';
```

**Comment by waynechen55** [comment]
@RoyCai7 After finalizing this pull request, please help do full verification run for both KVM and XEN.

File: lib/virt_autotest/utils.pm [hunk 1]
```diff
@@ -201,15 +201,25 @@ sub check_guest_health {
     return unless is_x86_64 and ($vm =~ /sle|alp/i);
 
     #check if guest is still alive
-    validate_script_output("virsh domstate $vm", sub { /running/ }, proceed_on_failure => 1) if (script_run("virsh list --all | grep \"$vm \"") == 0);
+    my $vmstate = "nok";
+    my $failures = "";
+    if (script_run("virsh list --all | grep \"$vm \"") == 0) {
+        $vmstate = "ok" if (script_run("virsh list --all | grep \"$vm \" | grep running") == 0);
```
Comment (waynechen55) [inline lib/virt_autotest/utils.pm:207]: How about change this line to 
```
$vmstate = "ok" if (script_run("virsh domstate | grep running") == 0);
```

File: lib/virt_autotest/utils.pm [hunk 2]
```diff
@@ -201,15 +201,25 @@ sub check_guest_health {
...
+        script_retry("xl list $vm | grep \"-b----\"", delay => 10, retry => 1, die => 0) for (0 .. 3);
+        $vmstate = "ok" if script_run("xl list $vm | grep \"-b----\"");
+    }
+    if ($vmstate eq "ok") {
+        $failures = check_failures_in_journal($vm);
+        return 'fail' if $failures;
+        record_info("Healthy guest!", "$vm looks good so far!");
+    }
+    else {
+        record_info("Skip check_failures_in_journal for $vm", "$vm is not active judged by either virsh or xl tool stack");
```
Comment (waynechen55) [inline lib/virt_autotest/utils.pm:219]: How about change this line to
```
record_info("Skip check_failures_in_journal for $vm", "$vm is not in desired state judged by either virsh or xl tool stack");
```

File: lib/virt_autotest/utils.pm (hunk 1 above)
Comment (RoyCai7) [reply lib/virt_autotest/utils.pm:207]: The same meaning, the new one is much better.

File: lib/virt_autotest/utils.pm (hunk 2 above)
Comment (RoyCai7) [reply lib/virt_autotest/utils.pm:219]: Agree with you, more accurate.

**Comment by waynechen55** [comment]
Some verification runs:
[15sp5 on 15sp5 kvm](https://openqa.suse.de/tests/10984929)
[15sp5 on 15sp5 xen](https://openqa.suse.de/tests/10984930)
[15sp5 on 12sp5](https://openqa.suse.de/tests/10984973)
[12sp5 on 15sp5](https://openqa.suse.de/tests/10984932)

**Review by alice-suse — APPROVED** [review]
LGTM

**Pull request merged**